[mod] set 'engine.supported_languages' from the origin python module

return42 opened the pull request:
## What does this PR do?

Set `engine.supported_languages` from the origin python module

The key of the dictionary 'searx.data.ENGINES_LANGUAGES' is the *engine name*
configured in settings.xml.  When multiple engines are configured to use the
same origin engine (e.g. `engine: google`)::

    - name: google
      engine: google
      use_mobile_ui: false
      ...

    - name: google italian
      engine: google
      use_mobile_ui: false
      language: it
      ...

    - name: google mobile ui
      engine: google
      shortcut: gomui
      use_mobile_ui: true

There exists no entry for ENGINES_LANGUAGES[engine.name] (e.g. `name: google
mobile ui` or `name: google italian`).  This issue can be solved by recreate the
ENGINES_LANGUAGES::

    make data.languages

But this is nothing an SearXNG admin would like to do when just configuring
additional engines, since this just doubles entries in ENGINES_LANGUAGES and
BTW: `make data.languages` has various external requirements which might be not
installed or not available, on a production host.

With this patch, if engine.name fails, ENGINES_LANGUAGES[engine.engine] is used
to get the engine.supported_languages (e.g. `google` for the engine named
`google mobile`).

For an engine, when there is `language: ...` in the YAML settings, the engine
supports only one language, in this case engine.supported_languages should
contains this value defined in settings.yml (e.g. `it` for the engine named
`google italian`).

## How to test this PR locally?

Activate the engines shown above in the settings.yml, call `make run` and do some queries e.g.  `!goit foo`, `!goit foo :en`, `!gomui foo :en`, `!gomui foo :it` and `foo`

## Related issues

Closes: https://github.com/searxng/searxng/issues/384

dalf commented on searx/engines/__init__.py:
```diff
@@ -149,8 +149,13 @@ def update_engine_attributes(engine, engine_data):
         engine.supported_languages = ENGINES_LANGUAGES[engine.name]
+    else:
+        # search for the configured engine from YAML setting: ``engine: ...``
+        if engine.engine in ENGINES_LANGUAGES:
```
Nitpicking, lines 155 to 157 can be written:
```python
elif engine.engine in ENGINES_LANGUAGES:
```

return42 replied:
OK, I will fix .. another question, instead of `engine.engine` we can also use `engine.__name__` what do you think is better?

dalf replied:
`engine.__name__` is related to https://github.com/searxng/searxng/issues/312

return42 replied:
> `engine.__name__` is related to https://github.com/searxng/searxng/issues/312

not really .. #312 is talking about registering modules in sys.modules, when loaded by `searx.utils.load_module(filename, module_dir)` .. the `engine.__name__` is set by `importlib` anyway .. and is not related to sys.modules.

dalf replied:
```python
import searx
import searx.engines.google
print(searx.engines.google.__name___)
# searx.engines.google
searx.engines.load_engines(searx.settings['engines'])
print(searx.engines.engines['google'].__name__)
# google
```

Is it expected (?)
I _guess_ the reason of the second output is (?):
https://github.com/searxng/searxng/blob/feb2d4dda575eb1501fb5a8427de364bed7bc179/searx/utils.py#L434
where `modname == 'google'`

return42 replied:
> I guess the reason of the second output is (?)

I think so, importlib sets the `__name__` when the module is loaded  ... don't ask me any details, the "importlib" is mystic to me ... but you can repeat your example with this patch applied and the result is the same ... this was also what I have tested before I posted https://github.com/searxng/searxng/pull/386#discussion_r723089965

```diff
diff --git a/searx/utils.py b/searx/utils.py
index 22824d829..c9ec0bc8d 100644
--- a/searx/utils.py
+++ b/searx/utils.py
@@ -427,13 +427,10 @@ def match_language(locale_code, lang_list=[], custom_aliases={}, fallback='en-US
 
 def load_module(filename, module_dir):
     modname = splitext(filename)[0]
-    if modname in sys.modules:
-        del sys.modules[modname]
     filepath = join(module_dir, filename)
     # and https://docs.python.org/3/library/importlib.html#importing-a-source-file-directly
     spec = importlib.util.spec_from_file_location(modname, filepath)
     module = importlib.util.module_from_spec(spec)
-    sys.modules[modname] = module
     spec.loader.exec_module(module)
     return module
 ```

> Is it expected (?)

I would expect, that the importlib stets the `__name__` whenever a module is loaded. This is at least what Python expect:

[ Within a module, the module’s name (as a string) is available as the value of the global variable `__name__`](https://docs.python.org/3/tutorial/modules.html)

~BTW: do you like to send a PR that applies the patch above (copied from your https://github.com/searxng/searxng/issues/312)~ .. see #390

return42 committed:
[mod] set 'engine.supported_languages' from the origin python module

The key of the dictionary 'searx.data.ENGINES_LANGUAGES' is the *engine name*
configured in settings.xml.  When multiple engines are configured to use the
same origin engine (e.g. `engine: google`)::

    - name: google
      engine: google
      use_mobile_ui: false
      ...

    - name: google italian
      engine: google
      use_mobile_ui: false
      language: it
      ...

    - name: google mobile ui
      engine: google
      shortcut: gomui
      use_mobile_ui: true

There exists no entry for ENGINES_LANGUAGES[engine.name] (e.g. `name: google
mobile ui` or `name: google italian`).  This issue can be solved by recreate the
ENGINES_LANGUAGES::

    make data.languages

But this is nothing an SearXNG admin would like to do when just configuring
additional engines, since this just doubles entries in ENGINES_LANGUAGES and
BTW: `make data.languages` has various external requirements which might be not
installed or not available, on a production host.

With this patch, if engine.name fails, ENGINES_LANGUAGES[engine.engine] is used
to get the engine.supported_languages (e.g. `google` for the engine named
`google mobile`).

For an engine, when there is `language: ...` in the YAML settings, the engine
supports only one language, in this case engine.supported_languages should
contains this value defined in settings.yml (e.g. `it` for the engine named
`google italian`).

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>
Closes: https://github.com/searxng/searxng/issues/384

dalf commented on searx/engines/__init__.py:
```diff
@@ -146,11 +146,29 @@ def update_engine_attributes(engine, engine_data):
+    # For an engine, when there is `language: ...` in the YAML settings, the
+    # engine supports only one language, in this case engine.supported_languages
+    # should contains this value defined in settings.yml
+    language = engine_data.get('language', None)
```
`engine_data.get('language')` since `None` is implicit.

dalf commented on searx/engines/__init__.py:
```diff
@@ -146,11 +146,29 @@ def update_engine_attributes(engine, engine_data):
+    language = engine_data.get('language', None)
+    if language:
+        if isinstance(engine.supported_languages, dict):
+            engine.supported_languages = {language : engine.supported_languages[language]}
```
This code crashes when `engine.supported_languages[language]` doesn't exist and `engine.supported_languages` is dictionary.

So it crashes with (see [engine_languages.json](https://github.com/searxng/searxng/blob/feb2d4dda575eb1501fb5a8427de364bed7bc179/searx/data/engines_languages.json#L25446-L25451) ) :
```yaml
- engine: google
   language: zh-Hans-CN
   ...
```

But not with (see [engine_languages.json](https://github.com/searxng/searxng/blob/feb2d4dda575eb1501fb5a8427de364bed7bc179/searx/data/engines_languages.json#L237-L240) ):
```yaml
- engine: bing
   language: zh-Hans-CN
   ...
```

I suggest to replace line 168 by:
```python
if isinstance(engine.supported_languages, dict) and language in engine.supported_languages:
```

return42 replied:
> This code crashes ..

I know, IMO it is a miss configuration when the configured language is not in the supported languages .. I also would like to add an exception when language is not in a list type:
```python
else:
    if language not in engine.supported_languages:
        raise KeyError(language)
    engine.supported_languages = [language] 
```

I looked  for an Exception-Type that fits better, but haven't found one .. may be should use SearxException?

https://github.com/searxng/searxng/blob/adeb084cf4ffc59f1e49b47d049e7541be6dd5dd/searx/exceptions.py#L19-L20

dalf replied:
`SearxSettingsException` ?

https://github.com/searxng/searxng/blob/master/searx/settings_loader.py#L22-L29

return42 replied:
This is what I thought first, but this exception type is only for loading the YAML fie and needs a `file_name` argument

    SearxSettingsException(e, file_name) f

dalf replied:
...

https://github.com/searxng/searxng/blob/adeb084cf4ffc59f1e49b47d049e7541be6dd5dd/searx/settings_defaults.py#L116-L132

`settings_defaults`:
* log details using `logger`
* raises a `ValueError` exception.

dalf commented on searx/engines/__init__.py:
```diff
@@ -146,11 +146,29 @@ def update_engine_attributes(engine, engine_data):
+        # settings.yml to use the same origin engine (python module) these
+        # additional engines can use the languages from the origin engine.
+        # For this use the configured ``engine: ...`` from settings.yml
+        engine.supported_languages = ENGINES_LANGUAGES[engine.engine]
```
Nitpicking / optional / not important, it can be :
```python
engine.supported_languages = ENGINES_LANGUAGES[engine_data["engine"]]
```

so the change on line 140 is not necessary.

return42 replied:
I can fix it like you suggested .. but when I implemented 140, I looked back in the commit history and did not found a reason why `elif param_name != 'engine'` is better than a simple `else` .. IMO the `elif` is a relic from the old days which is not needed anymore.

dalf replied:
I think the reason is redundancy `engine.__name__ == engine.engine`.

But as described, nitpicking.

dalf reviewed (changes requested):
I've added some comments.

---

Yes  `searx.engines.__init__` is the right location to initialize `engine.supported_languages` (Sorry, my comment was not clear).

dalf reviewed (approved):
Thank you!
